v2/api/authorization/customizations/role_assignment_extensions.go

```diff
@@ -43,3 +55,112 @@ func (extension *RoleAssignmentExtension) Import(
 
 	return result, nil
 }
+
+var _ extensions.ARMResourceModifier = &RoleAssignmentExtension{}
+
+func (extension *RoleAssignmentExtension) ModifyARMResource(
+	ctx context.Context,
+	armClient *genericarmclient.GenericClient,
+	armObj genruntime.ARMResource,
+	obj genruntime.ARMMetaObject,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) (genruntime.ARMResource, error) {
+	ra, ok := obj.(*storage.RoleAssignment)
+	if !ok {
+		return nil, eris.Errorf(
+			"Cannot run RoleAssignmentExtension.ModifyARMResource() with unexpected resource type %T",
+			obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not been updated to match
+	var _ conversion.Hub = ra
+
+	// If the specified role definition uses a well known name, look it up
+	roleDefinitionName := ra.Spec.RoleDefinitionReference.WellKnownName
+	if roleDefinitionName != "" {
+		err := ensureBuiltInRoleDefinitionsLoaded(ctx, armClient)
+		if err != nil {
+			return nil, eris.Wrapf(err, "loading built in role definitions to resolve %q", roleDefinitionName)
+		}
+
+		roleDefinitionId, err := resolveBuiltInRoleDefinition(roleDefinitionName, armObj)
+		if err != nil {
+			return nil, eris.Wrapf(err, "resolving built in role definition %q", roleDefinitionName)
+		}
+
+		if roleDefinitionId != roleDefinitionName {
+			log.V(1).Info("Resolved built-in role", "roleName", roleDefinitionName, "roleId", roleDefinitionId)
+
+			spec := armObj.Spec()
+			err = reflecthelpers.SetProperty(spec, "Properties.RoleDefinitionId", &roleDefinitionId)
+			if err != nil {
+				return nil, eris.Wrapf(err, "error setting RoleDefinitionId to %s", roleDefinitionId)
+			}
+
+			return armObj, nil
+		}
+	}
+
+	return armObj, nil
+}
+
+var builtInRoleDefinitions map[string]string = make(map[string]string, 800)
```

Comment:
Parallelism risk here. We can have many callers of this. Map is constructed once and written to many times. Go maps are not concurrent-safe.

Additionally, _technically speaking_ it's possible for new role definitions to be added. Depending on how long the ASO pod runs for, that could happen during its lifetime, but if we only ever write it once they won't be able to use the new ones without a pod restart.

We could have a threadsafe cache (there are some existing OSS implementations, Karpenter uses patrickmn/go-cache, though it doesn't have generics support and I think there are others that do) with a really long expiry like 1d?

Reply:
Cache is maybe not required (I'm open to either opinion). Thread-safety is required I think.

Reply:
very minor: Possibly worth a note in a comment here just acknowledging this case?